### ISSUE TITLE
exp: train contacts-and-coordinates-v1 1.5B on TPU (v5p), mixed with contacts-v1, from Eric #75

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#121]
## Goal

Train a 1.5B protein language model on the **contacts-and-coordinates-v1** corpus (exp105, #105) **mixed with the contacts-v1 corpus** (exp53), reusing **Eric's #75 winning recipe** (the tuned 1.47B Qwen3, `eval/contacts-v1-val/loss = 2.7566` @ 8 epochs), on **TPU (v5p) on the `marin` iris cluster**.

- **Mix:** 1 Å coordinate docs (bulk) + contacts-v1 docs (a token-minority). Train **4 epochs of each dataset** (see the mix table — this makes contacts-v1 ~3.3% of tokens).
- **Track validation loss on each dataset separately** (`eval/ccoord-v1/loss` and `eval/contacts-v1/loss`).

Same proteins/rounds/splits across both corpora, so this cleanly tests whether adding 1 Å coordinates to the document helps, while the contacts-v1 mix-in keeps the pure-contacts capability sharp and gives a metric directly comparable to the contacts-v1 models (#75 / exp67 / exp108 / exp112).

## Model — Eric #75 config, verbatim (do not change the architecture)

Qwen3 **1.47B** — use levanter `Qwen3Config` (as in exp108) on the TPU `default_train` path (exp67/85):

- `hidden_dim=2048`, `intermediate_dim=8192`, `num_layers=24`, `num_heads=32`, `num_kv_heads=8` (GQA 4:1), `head_dim=64`
- `use_qk_norm=True`, RMSNorm, SwiGLU, untied embeddings, `rope=Llama3RotaryEmbeddingsConfig()` (theta 500000)
- Optimizer: AdamW, `wd=0.2` (exclude norms + embeddings/output from decay), betas `0.9/0.95`, cosine (`min_lr_ratio=0.1`), **warmup 0.1**, no z-loss

> Model size held at #75's 1.5B for comparability. (Aside: ~560 B training tokens is far past Chinchilla-optimal for 1.5B — a bigger model is a reasonable follow-up, out of scope here.)

## Data — mix contacts-and-coordinates-v1 + contacts-v1 (one tokenizer, GCS-native)

Both corpora are already on **GCS** (TPU jobs read GCS directly — **no S3 / CoreWeave / Megatron `.bin` staging**, unlike the GPU plan). Levanter tokenizes on the fly and builds a token cache on GCS.

| corpus | GCS path (`{train,val}/*.parquet`, col `document`) | train docs | train tokens |
|---|---|--:|--:|
| ccoord-v1 (exp105) | `gs://marin-us-central1/protein-structure/MarinFold/exp105_ccoord_v1/documents/` | 4,129,682 | 135.3 B |
| contacts-v1 (exp53) | `gs://marin-us-east5/protein-structure/MarinFold/exp53_contacts_v1_5x/documents/` | 4,129,682 | 4.67 B |

**One tokenizer for both: `open-athena/contacts-and-coordinates-v1-tokenizer` (3,847 vocab).** Verified: contacts-v1's 2,845-token vocab is a **strict subset with identical token→id mappings** (0 missing, 0 remapped; both WordLevel), so contacts-v1 docs tokenize identically under it. (The 1,002 extra coordinate tokens never appear in contacts-v1 docs, so the contacts-v1 val loss stays ~comparable to #75's 2.7566 — only a negligible softmax-normalization difference.)

**Mixing (levanter `LmDataConfig`):** two `DatasetComponent`s in `components`, each with its `validation` split; set `train_weights` **∝ each corpus's token count** so both complete **4 epochs together**:

| | train_weights | docs | tokens | epochs |
|---|--:|--:|--:|--:|
| ccoord-v1 | **≈ 0.967** | 16.5 M passes | 541.2 B | 4 |
| contacts-v1 | **≈ 0.033** | 16.5 M passes | 18.7 B | 4 |
| **total** | 1.0 | 33.0 M | **≈ 560 B** | 4 |

Why ∝ token size and not 1:1: levanter's mixture weight is **per packed sequence** (≈ per token at seq 32768). Because contacts-v1 docs are ~29× shorter (~1,131 vs 32,766 tokens) they pack ~29 docs/sequence, so a **1:1 weight would give 50/50 tokens** (contacts-v1 *not* a minority) and ~29× unequal document passes. Weighting ∝ corpus token size yields the intended outcome: **documents 1:1 (equal mix), contacts-v1 = 3.3% of tokens (the minority), 4 epochs of each.** `pack=True`, unmasked loss, `block_cross_document_attention=True`, shuffle (Feistel), `data_seed=0`.

**Per-dataset eval:** give each component a distinct dict key + a `validation` split; levanter appends the component key as a tag → **`eval/ccoord-v1/loss`** and **`eval/contacts-v1/loss`** logged separately (plus aggregate `eval/loss` micro-avg and `eval/macro_loss`). Full val (`max_eval_batches=None`), unmasked.

**Region:** the two corpora are in **different GCS regions** (ccoord-v1 = `us-central1`, contacts-v1 = `us-east5`). Pin the TPU zone to **`us-central1-a`** (co-locate with the 303 GiB ccoord bulk + the checkpoint/token-cache prefix) and **mirror the small contacts-v1 corpus (~12 GiB) into `us-central1`** to avoid cross-region I/O on every cache build/checkpoint. (exp67/85 pinned `us-east5-a` for the opposite reason; flip it here.)

## Compute — TPU v5p on the `marin` iris cluster

Launch via the marin executor (the exp67/85 TPU path), not the exp108 GPU direct-dispatch:

```
uv run iris --cluster marin job run --no-wait --enable-extra-resources \
  --cpu=1 --memory=16GB --disk=16GB --extra=tpu --zone=us-central1-a -- python -m <train_module>
```

- **Resources:** `ResourceConfig.with_tpu("v5p-128", slice_count=1, cpu=32, ram="128g", disk="50g", zone="us-central1-a")` passed as `SimpleTrainConfig.resources` → `default_train(...)` → `TrainLmOnPodConfig`.
- **Recommended size: v5p-128** (128 chips ≈ 64×H100 in peak bf16 FLOPs — matches the originally-requested GPU scale). Tunable: v5p-256 to ~halve wall-clock; `slice_count>1` for multi-slice (DCN).
- **No multi-node ceiling here.** JAX-on-TPU multi-slice scales natively via the built-in coordinator (the GPU 8-node JAX-bootstrap failure that blocked the rno-2a plan does **not** apply to TPU). **NeMo is dropped** (CUDA-only). Framework = **Levanter/JAX**.
- Rough estimate: ~560 B tokens, 1.5B model → ~2–4 days on **v5p-128**, ~1–2 days on **v5p-256** (long context lowers MFU somewhat). Validate on the smoke run.

## What to scale, and how

| knob | #75 / exp67 | this run (recommended) | rationale |
|---|---|---|---|
| `seq_len` | 8,192 | **32,768** | native context of ccoord-v1; contacts-v1 docs pack into it |
| global batch | 128 seq → 1.05 M tok/step | **256 seq → 8.39 M tok/step** (fits v5p-128 at ~2 seq/chip; tunable) | keep the pod fed; sets step count |
| `num_train_steps` | — | **≈ 66,750** (560 B / 8.39 M) | 4 epochs of the mix |
| learning rate | 1e-3 | **1e-3 (anchor) → up to ~2e-3** | batch ~8× larger in tokens → √-scaling ≈ 2.8e-3; start safe. Single run, not a sweep, unless you want one |
| warmup | 10% | 10% | |
| grad checkpointing | on | **on (mandatory)** | 4× activation memory at seq 32768 |
| attention | — | **flash/splash (TPU), never vanilla** | seq 32768 → 16× score matrix; vanilla OOMs |

**Revisit RoPE for native 32768.** #75's Llama3 rope (theta 500000, scaling orig=8192, factor 8) was tuned for 8192; we now train natively at 32768 — sanity-check the rope base / Llama3 scaling params for 32768 positions rather than inheriting the 8192 settings blindly.

## Path / templates

- **TPU launch + `default_train`/`SimpleTrainConfig`:** exp67 (`experiments/exp67_models_contacts_v1_1_5b/{train_protein_1_5b_contacts_v1.py, contacts_v1_train_common.py}`) and exp85 (reheat). `ResourceConfig.with_tpu(...)`, `--extra=tpu`, marin executor.
- **#75 Qwen3 model + LR/wd/warmup recipe + mixture/eval `build_data_config` shape:** exp108 (`experiments/exp108_models_qwen_3b_contacts_v1/{dispatch_train.py, contacts_v1_train_common.py}`, worktree `musing-euclid-fd5ea5`).
- **levanter mixing + per-tag eval:** `marin/lib/levanter/src/levanter/data/text/datasets.py` (`LmDataConfig.components` / `train_weights` / `DatasetComponent`), `.../data/mixture.py` (weights per sequence), `.../eval.py` (`eval/<tag>/loss`).
- **models/ builder (main, TPU):** `models/marinfold_models/{defaults.py (default_train), simple_train_config.py}`. Always **commit a `uv.lock`** (pin marin) — see the exp108/exp85 marin-drift lesson.

## Success criteria

1. A converged 1.5B model trained with #75's recipe on the mixed corpus; TPU on `marin`, checkpoints to GCS (`us-central1`), W&B (`MarinFold` project).
2. **Both val losses tracked separately:** `eval/ccoord-v1/loss` (primary) and `eval/contacts-v1/loss` (secondary; ~comparable to #75's 2.7566 as a sanity signal).
3. Smoke run first (≤100 steps, real mixed-data path, both caches build, checkpoint) before the full run.
4. Report measured s/step, tok/s, MFU, TPU size; HF export of the final checkpoint (co-locate the ccoord-v1 tokenizer).
5. Confirm the mix behaves as intended (per-source example/token counts ≈ the table; both hit 4 epochs together).

## Open decisions to confirm before the full run

- **TPU size:** v5p-128 (≈ 64 H100, ~2–4 days) vs v5p-256 (~1–2 days)?
- **LR:** anchor 1e-3, or ~2e-3 for the 8× batch (or a 2-point {1e-3, 2e-3} probe)?
- **Token budget:** confirm **4 epochs of each** (≈ 560 B tokens; note this is heavily over-trained for 1.5B — intentional?).
- **Region:** mirror contacts-v1 into `us-central1` (recommended) vs read it cross-region from `us-east5`.

## References

- Datasets: exp105 / #105 (ccoord-v1, PR #119); exp53 (contacts-v1)
- TPU training template: exp67 / #67, exp85 (`experiments/exp67_models_contacts_v1_1_5b/`)
- #75 Qwen3 recipe + mixture/eval shape: exp108 / #108, PR #111; #75 = marin `eac/plm-exp75`, W&B `eric-czech/marin`, best `prot-exp75-cv1-1_5b-e8-lr1e-3-wd0p2` (eval loss 2.7566)
- levanter data/mixing/eval: `marin/lib/levanter/src/levanter/{data/text/datasets.py, data/mixture.py, eval.py}`

---

_Per the MarinFold convention, the experiment number = this issue number (#121); use it for the branch, `experiments/exp121_.../` dir, and run names._

_(Revised from the original GPU/rno-2a draft: switched to TPU v5p on `marin`, added the contacts-v1 mix-in with per-dataset eval, dropped all S3/CoreWeave/NeMo/`.bin` and the GPU multi-node-ceiling content.)_

**Comment by timodonnell** [comment]
Superseded by **contacts-and-crops-v1** — closing in favor of a separate training experiment on that format.

**Why.** This issue proposed training a 1.5B on the contacts-and-coordinates-v1 (ccoord) corpus mixed with contacts-v1. Since it was filed, [#130](https://github.com/Open-Athena/MarinFold/issues/130) / PR [#131](https://github.com/Open-Athena/MarinFold/pull/131) introduced **contacts-and-crops-v1**, which carries coordinates in an **8192-token** context instead of ccoord's 32768 — coarse 10 Å boxes for every atom (Pass 1) plus 0.1 Å detail inside a handful of `<crop>` boxes (Pass 2). That makes it strictly preferable as the first coordinate-training target:

- **~4x cheaper context** — trains at contacts-v1 cost rather than 32k-context cost.
- **Mixes 1:1 with the contacts formats** at the same sequence length, so no ragged mixing.
- **Warm-starts from a contacts-v1 / ccoord checkpoint** — only two new tokens (`<contacts-and-crops-v1>`, `<crop>`); the `<xyz-DDD>` vocab is shared.

**Data status.** Both corpora are generated and published on the `open-athena/MarinFold` bucket over the same afdb-24M entries, clusters, pLDDT rounds, and splits:

- `data/document_structures/contacts_and_crops_v1/` — 4,213,203 docs, ~34.5B tokens ([#132](https://github.com/Open-Athena/MarinFold/issues/132))
- `data/document_structures/contacts_and_coordinates_v1/` — same 4.21M entries at 32k ([#105](https://github.com/Open-Athena/MarinFold/issues/105))

So no data work is lost here: the ccoord corpus stays on the bucket, and if we later want a **32k-vs-8k coordinate-resolution comparison** it can be run as its own experiment against this same recipe.

**Next.** A separate issue will cover training a 1.5B on contacts-and-crops-v1 (mixed with contacts-v1, warm-started from Eric's [#75](https://github.com/Open-Athena/MarinFold/issues/75) recipe) — the recipe, mix table, and per-dataset validation-loss tracking described here carry over essentially unchanged.
